File: Opentutorials/py_rb/override/2.py
class Cal(object):
    _history = []
    def __init__(self, v1, v2):
        if isinstance(v1, int):
            self.v1 = v1
        if isinstance(v2, int):
            self.v2 = v2
    def add(self):
        result = self.v1 + self.v2
        Cal._history.append("add : %d + %d= %d" %(self.v1, self.v2, result))
        return result
    def sub(self):
        result = self.v1 - self.v2
        Cal._history.append("add : %d - %d= %d" %(self.v1, self.v2, result))
        return result
    def setV1(self, v):
        if isinstance(v, int):
            self.v1 = v
    def getV1(self):
        return self.v1
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)
    def info(self):
        return "Cal => v1 : %d, v2 : %d" % (self.v1, self.v2)
class CalMulti(Cal):
    def multi(self):
        result = self.v1 * self.v2
        Cal._history.append("add : %d * %d= %d" %(self.v1, self.v2, result))
        return result
    def info(self):
        return "calMultiply=> %s" %super().info()
class CalDiv(CalMulti):
    def devi(self):
        result = self.v1 / self.v2
        Cal._history.append("add : %d / %d= %d" %(self.v1, self.v2, result))
        return result
    def info(self):
        return "calMultiply=> %s" %super().info()
c0 = Cal(30, 60)
print(c0.info())

c1 = CalMulti(10,10)
print(c1.info())
c2 = CalDiv(10,10)
print(c2.info())

File: Opentutorials/py_rb/override/2.rb
class Cal
    attr_reader : v1, v2
    attr_writer : v1
    @@_history = []
    def initialize(v1, v2)
        @v1 = v1
        @v2 = v2
    end
    def add()
    result = @v1 + @v2
    @@_history.push("add : #{@v1} + #{@v2} = #{result}")
    return result
    end
    def setV1(v)
        if v.is_a?(Integer)
            @v1 = v
        end
    end
    def getV1()
        return @v1
    end
    def Cal.history()
        for item in @@_history
            p item
        end
    end
    def info()
        return "Cal => v1 : #{@v1}, v2 : #{@v2}" 
    end
end
class CalMultiply < Cal
    def multiply()
        result = @v1 *@v2
        @@_history.push("multifly : #{@v1} * #{@v2} = #{result}")
        return result
    end
    def info()
        return "CalMultiply => v1 : #{@v1}, v2 : #{@v2}" 
    end
end
class CalDivide < CalMultiply
    def devide()
        result = @v1 / @v2
        @@_history.push("multifly : #{@v1} * #{@v2} = #{result}")
        return result
    end
    def info()
        return "CalDivide => v1 : #{@v1}, v2 : #{@v2}" 
    end
end

c0 =Cal.new(30, 60)

c1 = CalMultiply.new(10,10)
p c1.add()
p c1.multifly()
c2 = CalDivide.new(10, 10)